src/utils.js

```
@@ -314,26 +314,84 @@ export function optionsList(schema) {
-  if (match && match[1]) {
-    const parts = match[1].split("/");
-    let current = definitions;
-    for (let part of parts) {
-      part = part.replace(/~1/g, "/").replace(/~0/g, "~");
-      if (current.hasOwnProperty(part)) {
-        current = current[part];
-      } else {
-        // No matching definition found, that's an error (bogus schema?)
-        throw new Error(`Could not find a definition for ${$ref}.`);
+  if (isJSONPointer($ref)) {
+    const match = /^#\/definitions\/(.*)$/.exec($ref);
+    if (match && match[1]) {
+      const parts = match[1].split("/");
+      let current = definitions;
+      for (let part of parts) {
+        part = part.replace(/~1/g, "/").replace(/~0/g, "~");
+        if (current.hasOwnProperty(part)) {
+          current = current[part];
+        } else {
+          // No matching definition found, that's an error (bogus schema?)
+          throw new Error(`Could not find a definition for ${$ref}.`);
+        }
       }
+      return current;
+    }
+  } else if (is$id($ref)) {
+    if (has$id($ref, definitions)) {
+      let keys = Object.keys(definitions);
+      for (let key of keys) {
+        let { $id: key$id } = definitions[key];
+        if (key$id === $ref) {
+          let current = definitions[key];
+          if (current.hasOwnProperty("$ref") && is$id(current.$ref)) {
+            processedIds = processedIds.concat($ref);
+            //One question refers to itself or to another question inside it, used for object and array schemas
+            if (processedIds.includes(current.$ref)) {
+              throw new Error(
+                `Circular or self reference detected when reading schema for ${current.$ref}`
+              );
+            }
+            return findSchemaDefinition(
+              current.$ref,
+              definitions,
+              processedIds
+            );
+          }
+          return current;
+        }
+      }
+    } else {
```

I have added more cases to flesh out where errors are thrown, I think it should throw an Error only when an illegal item is found.